File: KrkrDumpBootstrap/bootstrap/CxdecOrder.c
//----- (654E8610) --------------------------------------------------------
char __thiscall sub_654E8610(int this, const char * a2) {
    v8 = cxrandom & 7;
    if (v8 ==  * v7) // 0
        result = sub_654E8B60((_DWORD * )this, 0xF7, 0xD0);
    if (v8 == v7[1]) // 2
        result = sub_654E8B60((_DWORD * )this, 0xF7, 0xD8);
    if (v8 == v7[2]) // 3
        result = sub_654E8B30((_DWORD * )this, 0x40);
    if (v8 == v7[3]) // 1
        result = sub_654E8B30((_DWORD * )this, 0x48);
    if (v8 == v7[4]) { // 5
        if (!sub_654E8B30((_DWORD * )this, 0x53)
             || !sub_654E8B60((_DWORD * )this, 0x89, 0xC3)
             || !sub_654E8C50((_DWORD * )this, 0x81, 0xE3, 0xAA, 0xAA, 0xAA, 0xAA)
             || !sub_654E8BF0((_DWORD * )this, 0x25, 0x55, 0x55, 0x55, 0x55)
             || !sub_654E8B60((_DWORD * )this, 0xD1, 0xEB)
             || !sub_654E8B60((_DWORD * )this, 0xD1, 0xE0)
             || !sub_654E8B60((_DWORD * )this, 9, 0xD8)) {
            return 0;
        }
    }
    if (v8 == v7[5]) // 6
        result = sub_654E8B30((_DWORD * )this, 0x35);
    if (v8 != v7[6]) { // 7
        result = sub_654E8B30((_DWORD * )this, 5);
        else if (!sub_654E8B30((_DWORD * )this, 0x2D))
    }
    if (v8 == v7[7]) { // 4
        if (!sub_654E8B30((_DWORD * )this, 0xBE)
             || !sub_654E8CC0((_DWORD * )this,  * (_DWORD * )(this + 0xC))
             || !sub_654E8B30((_DWORD * )this, 0x25)
             || !sub_654E8CC0((_DWORD * )this, 0x3FF)) {
            return 0;
        }
        result = sub_654E8BA0((_DWORD * )this, 0x8B, 4, 0x86);
    }
}

//----- (66048000) --------------------------------------------------------
char __thiscall sub_66048000(int this, const char * a2) {
    v11 = cxrandom % 6;
    v12 =  * (_BYTE ** )(this + 0x2C);
    if ((_BYTE)v11 == v12[8]) // 2
        result = sub_66048B60((_DWORD * )this, 0x01, 0xD8);
    if ((_BYTE)v11 == v12[9]) // 5
        result = sub_66048B60((_DWORD * )this, 0x29, 0xD8);
    if ((_BYTE)v11 != v12[0xA]) { // 3
        if (!sub_66048B60((_DWORD * )this, 0xF7, 0xD8))
            result = sub_66048B60((_DWORD * )this, 1, 0xD8);
    }
    if ((_BYTE)v11 == v12[0xB]) // 4
        result = sub_66048BA0((_DWORD * )this, 0xF, 0xAF, 0xC3);
    if ((_BYTE)v11 == v12[0xC]) { // 1
        if (!sub_66048B30((_DWORD * )this, 0x51)
             || !sub_66048B60((_DWORD * )this, 0x89, 0xD9)
             || !sub_66048BA0((_DWORD * )this, 0x83, 0xE1, 0xF)
             || !sub_66048B60((_DWORD * )this, 0xD3, 0xE0)) {
            return 0;
        }
        result = sub_66048B30((_DWORD * )this, 0x59);
    }
    if ((_BYTE)v11 == v12[0xD]) { // 0
        if (!sub_66048B30((_DWORD * )this, 0x51)
             || !sub_66048B60((_DWORD * )this, 0x89, 0xD9)
             || !sub_66048BA0((_DWORD * )this, 0x83, 0xE1, 0xF)
             || !sub_66048B60((_DWORD * )this, 0xD3, 0xE8)) {
            return 0;
        }
        result = sub_66048B30((_DWORD * )this, 0x59);
    }
}

//----- (10018410) --------------------------------------------------------
char __thiscall sub_10018410(int this) {
    v4 = cxrandom % 3;
    v5 =  * (_BYTE ** )(this + 0x2C);
    if (v4 == v5[0xE]) // 0
         * (_BYTE * )(v6 +  * (_DWORD * )this) = 0xB8;
    if (v4 == v5[0xF]) { // 1
         * (_BYTE * )(v13 +  * (_DWORD * )this) = 0x8B;
         * (_BYTE * )( * (_DWORD * )this +  * (_DWORD * )(this + 4) + 1) = 0xC7;
    }
    if (v4 == v5[0x10]) {// 2
        if (!sub_10018B30((_DWORD * )this, 0xBE))
            result = sub_10018B60((_DWORD * )this, 0x8B, 0x86);
        v17 = sub_10018B10((unsigned int * )this) & 0x3FF;
        result = sub_10018CC0((_DWORD * )this, 4 * v17);
    }
}

File: KrkrDumpBootstrap/bootstrap/CxdecOrder.js
const fs = require("fs");

const garbro_evenmap = [
    "NOT_EAX",
    // NOT_EAX 0xf7 0xd0

    "DEC_EAX",
    // DEC_EAX 0x48

    "NEG_EAX",
    // NEG_EAX 0xf7 0xd8

    "INC_EAX",
    // INC_EAX 0x40

    "MOV_EAX_INDIRECT",
    // NOP              0xbe
    // AND_EAX_IMMED    0x25
    // 0x000003FF
    // MOV_EAX_INDIRECT 0x8b 0x04 0x86

    "INTERLACE_EAX",
    // PUSH_EBX      0x53
    // MOV_EBX_EAX   0x89 0xc3
    // AND_EBX_IMMED 0x81 0xe3
    // 0xAAAAAAAA
    // AND_EAX_IMMED 0x25
    // 0x55555555
    // SHR_EBX_1     0xd1 0xeb
    // SHL_EAX_1     0xd1 0xe0
    // OR_EAX_EBX    0x09 0xd8
    // POP_EBX       0x5b

    "XOR_EAX_IMMED",
    // XOR_EAX_IMMED 0x35

    "ADDORSUB_EAX_IMMED"
    // ADD_EAX_IMMED 0x05
    // SUB_EAX_IMMED 0x2d
];
const garbro_oddmap = [
    "SHR_EAX_CL",
    // PUSH_ECX    0x51
    // MOV_ECX_EBX 0x89 0xd9
    // AND_ECX_0F  0x83 0xe1 0x0f
    // SHR_EAX_CL  0xd3 0xe8
    // POP_ECX     0x59

    "SHL_EAX_CL",
    //PUSH_ECX    0x51
    //MOV_ECX_EBX 0x89 0xd9
    //AND_ECX_0F  0x83 0xe1 0x0f
    //SHL_EAX_CL  0xd3 0xe0
    //POP_ECX     0x59

    "ADD_EAX_EBX",
    //ADD_EAX_EBX 0x01 0xd8

    "NEG_EAX_ADD_EAX_EBX",
    // NEG_EAX     0xf7 0xd8
    // ADD_EAX_EBX 0x01 0xd8

    "IMUL_EAX_EBX",
    // IMUL_EAX_EBX 0x0f 0xaf 0xc3

    "SUB_EAX_EBX"
    // SUB_EAX_EBX 0x29 0xd8
];
const garbro_prologmap = [
    "MOV_EAX_IMMED",
    // MOV EAX, Random() 0xb8

    "MOV_EAX_EDI",
    // MOV_EAX_EDI 0x8b 0xc7

    "MOV_EAX_INDIRECT"
    // NOP 0xbe
    // MOV_EAX_IMMED 0x8b 0x86
    // GetRandom() & 0x000003FF
    // MOV_EAX_INDIRECT
];

const dump_evenmap = [];
dump_evenmap[0] = garbro_evenmap[0];
dump_evenmap[1] = garbro_evenmap[2];
dump_evenmap[2] = garbro_evenmap[3];
dump_evenmap[3] = garbro_evenmap[1];
dump_evenmap[4] = garbro_evenmap[5];
dump_evenmap[5] = garbro_evenmap[6];
dump_evenmap[6] = garbro_evenmap[7];
dump_evenmap[7] = garbro_evenmap[4];

const dump_oddmap = [];
dump_oddmap[0] = garbro_oddmap[2];
dump_oddmap[1] = garbro_oddmap[5];
dump_oddmap[2] = garbro_oddmap[3];
dump_oddmap[3] = garbro_oddmap[4];
dump_oddmap[4] = garbro_oddmap[1];
dump_oddmap[5] = garbro_oddmap[0];

const dump_prologmap = [];
dump_prologmap[0] = garbro_prologmap[0];
dump_prologmap[1] = garbro_prologmap[1];
dump_prologmap[2] = garbro_prologmap[2];

// console.log(garbro_evenmap)
// console.log(garbro_oddmap)
// console.log(garbro_prologmap)
// console.log(dump_evenmap)
// console.log(dump_oddmap)
// console.log(dump_prologmap)


let data = fs.readFileSync("CxdecOrder.bin");
let o1 = [];
let o2 = [];
let o3 = [];

for (let i = 0; i < 8; i++) {
	let r = data[i];
	let c = dump_evenmap[i];
	let x = garbro_evenmap.indexOf(c)
	o1[r]=x
	// console.log(r,c,x)
}
for (let i = 0; i < 6; i++) {
	let r = data[i+8];
	let c = dump_oddmap[i];
	let x = garbro_oddmap.indexOf(c)
	o2[r]=x
	// console.log(r,c,x)
}
for (let i = 0; i < 3; i++) {
	let r = data[i+14];
	let c = dump_prologmap[i];
	let x = garbro_prologmap.indexOf(c)
	o3[r]=x
	// console.log(r,c,x)
}

console.log(data)
console.log(o1.join(","))
console.log(o2.join(","))
console.log(o3.join(","))
